### ISSUE TITLE
Question: MNIST example test result on Wiki

### ISSUE THREAD

**Issue by pliptor** [tiny-dnn/tiny-dnn#566]
I'm sorry I'm not all familiar yet with everything but is it correct that the test result for the
classification for number 4 in the MNIST example on the Wiki page does not add up
to 100%? Please let me know if this is the correct behavior. There might be some normalization issue going on.

4,78.1403
7,33.5718
8,14.0017

I tried to replicate the test on my machine with the current head and I also get more than 100%

4,95.9938
7,8.09721
1,4.74199

![image](https://cloud.githubusercontent.com/assets/23116478/22849090/b44fc0be-efae-11e6-972f-a11740b789a1.png)

**Comment by pliptor** [comment]
Answering partially to my question, I think it has to do with not using a softmax function at the final layer. I'm now simulating the example code with tanh replanced by softmax to see the check it. In theory, I think it should give more meaningful outputs.

**Comment by pliptor** [comment]
I must be doing something wrong in trying to replace the tanh by softmax. The results look more off. If anyone has any insights please let me know.

4,112.158
7,50.1448
2,50.1084

**Comment by Randl** [comment]
Usually the output of network is interpreted as probability, but it isn't exactly. The closer output is to 1, the more probable the output is, but the output isn't the probability itself, that's why it don't have to add up to 1.

As for softmax, this seems really strange, because output of softmax is always smaller than 1.

**Comment by edgarriba** [comment]
You can interpret them with relative probabilities

**Comment by pliptor** [comment]
@Randl Let me dig a bit to see if I'm using softmax incorrectly.
@edgarriba  @Randl  Would it make more sense to present normalized outputs MNIST results? There might be something else going on because the way it is phrased on Wiki right now is mathematically questionable. "predicted this image as "4", at confidence level of 78.1403%": percentage infers probability.
If hypothetically one of the numbers is predicted with 100% confidence level, the others can't be anything else other than 0.

**Comment by pliptor** [comment]
I think I have an idea of what's going on so I'll write it here for documentation. There is a per output (not per vector) normalization attempt in the MNIST example test code (a scale function right at the top of the code). That is supposed to ensure each output is bounded between 0 and 100. the test code assumes, however, that the activation type is tan_h. If the user changes the training code to a different activation, e.g. softmax, the normalization may end up with a mismatch (this is especially true if tan_h is used).

If the above is correct, I think the nicest way to handle it would it be for the test code somehow follow the train code's last layer activation type so that the per output normalization is matched. Following that, a vector normalization would ensure that the classification presents numbers that a user would feel more intuitive.
